File: backend/games/GameOfLife.py
import numpy as np

class GameOfLife:
    def __init__(self, rows=10, columns=10):
        self.rows = rows
        self.columns = columns
        self.board = np.zeros((self.rows, self.columns), dtype=int)

    def reset(self, rows=None, columns=None):
        if rows is not None:
            self.rows = rows
        if columns is not None:
            self.columns = columns
        self.board = np.zeros((self.rows, self.columns), dtype=int)
        return {'board': self.board.tolist()}

    def set_cell(self, row, column, state):
        if 0 <= row < self.rows and 0 <= column < self.columns:
            self.board[row, column] = state
            return True
        return False

    def step(self):
        new_board = np.zeros((self.rows, self.columns), dtype=int)
        for row in range(self.rows):
            for column in range(self.columns):
                live_neighbors = self._count_live_neighbors(row, column)
                if self.board[row, column] == 1:
                    if live_neighbors < 2 or live_neighbors > 3:
                        new_board[row, column] = 0
                    else:
                        new_board[row, column] = 1
                else:
                    if live_neighbors == 3:
                        new_board[row, column] = 1
        self.board = new_board
        return {'board': self.board.tolist()}

    def _count_live_neighbors(self, row, column):
        neighbors = [
            (row-1, column-1), (row-1, column), (row-1, column+1),
            (row, column-1),                 (row, column+1),
            (row+1, column-1), (row+1, column), (row+1, column+1)
        ]
        count = 0
        for r, c in neighbors:
            if 0 <= r < self.rows and 0 <= c < self.columns:
                count += self.board[r, c]
        return count

File: frontend/src/games/GameOfLife/GameOfLife.js
import React, { useState, useEffect } from 'react';
import axios from 'axios';
import './GameOfLife.css';

const BASE_URL = 'http://localhost:5000';

const GameOfLife = () => {
    const [board, setBoard] = useState([]);
    const [rows, setRows] = useState(10); // Initial number of rows
    const [columns, setColumns] = useState(10); // Initial number of columns

    useEffect(() => {
        resetGame();
    }, []);

    const resetGame = async () => {
        try {
            const response = await axios.post(`${BASE_URL}/gameoflife-reset`, { rows, columns });
            setBoard(response.data.board);
        } catch (error) {
            console.error('Error resetting the game:', error);
        }
    };

    const setCell = async (row, column) => {
        try {
            const response = await axios.post(`${BASE_URL}/gameoflife-set`, { row, column });
            setBoard(response.data.board);
        } catch (error) {
            console.error('Error setting cell state:', error);
        }
    };

    const stepGame = async () => {
        try {
            const response = await axios.post(`${BASE_URL}/gameoflife-step`);
            setBoard(response.data.board);
        } catch (error) {
            console.error('Error advancing the game:', error);
        }
    };

    const renderBoard = () => {
        return board.map((row, rowIndex) => (
            <div key={rowIndex} className="row">
                {row.map((cell, colIndex) => (
                    <div
                        key={colIndex}
                        className={`cell ${cell ? 'alive' : 'dead'}`}
                        onClick={() => setCell(rowIndex, colIndex)}
                    ></div>
                ))}
            </div>
        ));
    };

    return (
        <div>
            <div>
                <div>
                    <label>
                        Rows:
                        <input
                            type="range"
                            min="1"
                            max="30"
                            value={rows}
                            onChange={(e) => setRows(parseInt(e.target.value))}
                        />
                    </label>
                    <label>
                        Columns:
                        <input
                            type="range"
                            min="1"
                            max="40"
                            value={columns}
                            onChange={(e) => setColumns(parseInt(e.target.value))}
                        />
                    </label>
                </div>
                <div style={{ display: 'flex' }}>
                    <button className='reset-button' onClick={resetGame}>Reset</button>
                    <button className='reset-button' onClick={stepGame}>Next Step</button>
                </div>
            </div>
            <div className="board">{renderBoard()}</div>
        </div>
    );
};

export default GameOfLife;
